lib/phoenix/socket.ex

```diff
@@ -37,27 +46,72 @@ defmodule Phoenix.Socket do
             topic: nil,
             transport: nil,
             transport_pid: nil
+end
+
+defmodule Phoenix.Socket.Message do
+  @moduledoc """
+  Defines a message dispatched over transport to channels and vice-versa.
+
+  The message format requires the following keys:
+
+    * `topic` - The string topic or topic:subtopic pair namespace, ie "messages", "messages:123"
+    * `event`- The string event name, ie "phx_join"
+    * `payload` - The message payload
+    * `ref` - The unique string ref
+
+  """
+
+  alias Phoenix.Socket.Message
+
+  defstruct topic: nil, event: nil, payload: nil, ref: nil
 
   @doc """
-  Sets topic of socket.
+  Converts a map with string keys into a message struct.
+
+  Raises `Phoenix.Socket.InvalidMessageError` if not valid.
   """
-  def put_topic(socket, topic) do
-    %Socket{socket | topic: topic}
+  def from_map!(map) when is_map(map) do
+    try do
+      %Message{
+        topic: Map.fetch!(map, "topic"),
+        event: Map.fetch!(map, "event"),
+        payload: Map.fetch!(map, "payload"),
+        ref: Map.fetch!(map, "ref")
+      }
+    rescue
+      err in [KeyError] ->
+        raise Phoenix.Socket.InvalidMessageError, message: "missing key #{inspect err.key}"
+    end
   end
+end
 
-  @doc """
-  Adds key/value pair to socket assigns.
+defmodule Phoenix.Socket.Reply do
+  @moduledoc """
+  Defines a reply sent from channels to transports.
 
-  ## Examples
+  The message format requires the following keys:
 
-      iex> socket.assigns[:token]
-      nil
-      iex> socket = assign(socket, :token, "bar")
-      iex> socket.assigns[:token]
-      "bar"
+    * `topic` - The string topic or topic:subtopic pair namespace, ie "messages", "messages:123"
+    * `status` - The reply status as an atom
+    * `payload` - The reply payload
+    * `ref` - The unique string ref
 
   """
-  def assign(socket = %Socket{}, key, value) do
-    put_in socket.assigns[key], value
-  end
+
+  defstruct topic: nil, status: nil, payload: nil, ref: nil
+end
+
+defmodule Phoenix.Socket.Broadcast do
```

Comment:
Besides Phoenix.Socket.Message, I have introduced Reply and Broadcast. Since we need to make this distinction in the test client, we may need to do it for other transports too. For example, if the client does a sync message, it will work now that we have specific Reply messages. It doesn't matter for JS because everything is async.

There is also Broadcast which is what goes into PubSub. I have created a specific struct for it because it may change in the future once we add things like Connectable. So better to set them apart now.